### PR TITLE
Null repository object

### PR DESCRIPTION
**Pull request by siemakuba**
This is a kind of solution for issue #188. 
It does not fix the race condition issue inside InMemoryRepository, but it adds a NullRepository.
It could be used in cases where RubyEventStore is used as a PubSub broker.

### CODE REVIEW

**Comment by paneq** [comment]
@siemakuba Thank you!, I will have a look once we handle #192 and #194 because it breaks our build currently and affects all PRs.

But just by looking at commit messages. We don't remove constants without deprecating them so moving `InMemoryRepository` to `Repositories::InMemory` is a breaking change for those who use it. My 2cents in the morning.

**Comment by siemakuba** [comment]
> We don't remove constants without deprecating them so moving InMemoryRepository to Repositories::InMemory is a breaking change for those who use it. 

Right, this is reasonable. I will adjust my PR accordingly. Thank you.

File: ruby_event_store/lib/ruby_event_store/repositories/null.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+      def method_missing(method_name, *_)
+        super unless respond_to_missing?(method_name)
+      end
+
+      def respond_to_missing?(method_name, include_private = false)
+        @repository_to_mimic.respond_to?(method_name, include_private)
+      end
+    end
+  end
+end
```
Comment (paneq) [inline ruby_event_store/lib/ruby_event_store/repositories/null.rb]: I had to check in IRB how would that work.

```
irb(main):007:0* module RubyEventStore
irb(main):008:1>   module Repositories
irb(main):009:2>     class Null
irb(main):010:3>       def initialize(repository_to_mimic = InMemory.new)
irb(main):011:4>         @repository_to_mimic = repository_to_mimic
irb(main):012:4>       end
irb(main):013:3> 
irb(main):014:3*       def method_missing(method_name, *_)
irb(main):015:4>         super unless respond_to_missing?(method_name)
irb(main):016:4>       end
irb(main):017:3> 
irb(main):018:3*       def respond_to_missing?(method_name, include_private = false)
irb(main):019:4>         @repository_to_mimic.respond_to?(method_name, include_private)
irb(main):020:4>       end
irb(main):021:3>     end
irb(main):022:2>   end
irb(main):023:1> end
=> :respond_to_missing?
irb(main):024:0> RubyEvetStore::Repositories::Null
NameError: uninitialized constant RubyEvetStore
	from (irb):24
	from /Users/rupert/.rbenv/versions/2.2.8/bin/irb:11:in `<main>'
irb(main):025:0> RubyEventStore::Repositories::Null
=> RubyEventStore::Repositories::Null
irb(main):026:0> RubyEventStore::Repositories::Null.new
=> #<RubyEventStore::Repositories::Null:0x007f90741ebf48 @repository_to_mimic=#<InMemory:0x007f90741ebf20>>
irb(main):027:0> RubyEventStore::Repositories::Null.new.asd
=> nil
irb(main):028:0> RubyEventStore::Repositories::Null.new.def
NoMethodError: undefined method `def' for #<RubyEventStore::Repositories::Null:0x007f9073832970>
	from (irb):15:in `method_missing'
	from (irb):28
	from /Users/rupert/.rbenv/versions/2.2.8/bin/irb:11:in `<main>'
```

Comment (paneq) [reply ruby_event_store/lib/ruby_event_store/repositories/null.rb]: I must say I am not comfortable with this solution. It's cleaver but here is what I would expect instead

* write methods (append_to_stream, link_to_stream, delete_stream) pass and return nil
* read methods (has_event?, read_events_forward, read_all_streams_backward, etc) raise custom exception

although that does not sound quite right as well

Comment (paneq) [reply ruby_event_store/lib/ruby_event_store/repositories/null.rb]: Technically we only need `append_to_stream` to work (do nothing, but exception-free) and everything else could rise an exception (which i prefer much more than `nil`).

**Comment by paneq** [comment]
@siemakuba I commented on some parts but I have deeper thoughts right now.

**Comment by paneq** [comment]
```ruby
module RubyEventStore
  class Client
    def initialize(repository:,
                   event_broker:  PubSub::Broker.new,
                   page_size: PAGE_SIZE,
                   metadata_proc: nil,
                   clock: ->{ Time.now.utc })
      @repository     = repository
      @event_broker   = event_broker
      @page_size      = page_size
      @metadata_proc  = metadata_proc
      @clock          = clock
    end


    def publish_events(events, stream_name: GLOBAL_STREAM, expected_version: :any)
      append_to_stream(events, stream_name: stream_name, expected_version: expected_version)
      events.each do |ev|
        @event_broker.notify_subscribers(ev)
      end
      :ok
    end

    def publish_event(event, stream_name: GLOBAL_STREAM, expected_version: :any)
      publish_events([event], stream_name: stream_name, expected_version: expected_version)
    end

    def append_to_stream(events, stream_name: GLOBAL_STREAM, expected_version: :any)
      events = normalize_to_array(events)
      events.each{|event| enrich_event_metadata(event) }
      @repository.append_to_stream(events, stream_name, expected_version)
      :ok
    end

    def subscribe(subscriber, event_types, &proc)
      @event_broker.add_subscriber(subscriber, event_types).tap do |unsub|
        handle_subscribe(unsub, &proc)
      end
    end

    def subscribe_to_all_events(subscriber, &proc)
      @event_broker.add_global_subscriber(subscriber).tap do |unsub|
        handle_subscribe(unsub, &proc)
      end
    end

    private

    def enrich_event_metadata(event)
      return unless event.respond_to?(:metadata)
      metadata = event.metadata
      metadata[:timestamp] ||= @clock.()
      metadata.merge!(@metadata_proc.call || {}) if @metadata_proc
    end

    def normalize_to_array(events)
      return *events
    end

    def handle_subscribe(unsub, &proc)
      begin
        proc.call
      ensure
        unsub.()
      end if proc
    end
```

Regarding #188

@andrzejkrzywda had an interesting thought that maybe if you are not using the repository this should tell us something deeper. Here is a thought: _you just want to use `PubSub::Broker.new`_ and not `RailsEventStore::Client`.  I guess you don't need `enrich_event_metadata` and probably we could move `handle_subscribe` into the `Broker`...

The API is different (subscribe -> add_subscriber, subscribe_to_all_events -> add_global_subscriber and most importantly publish_events -> notify_subscribers) but maybe we could make it similar.

So when I think about it...

Maybe it is better to create a custom wrapper on top of `PubSub::Broker.new` that would just do:

```
class PublishOnlyClient
    def initialize(event_broker:  PubSub::Broker.new)
      @event_broker   = event_broker
    end

    def publish_events(events, stream_name: nil, expected_version: nil)
      events.each do |ev|
        @event_broker.notify_subscribers(ev)
      end
      :ok
    end

    def publish_event(event, stream_name: nil, expected_version: nil)
      publish_events([event])
    end

    def subscribe(subscriber, event_types, &proc)
      @event_broker.add_subscriber(subscriber, event_types).tap do |unsub|
        handle_subscribe(unsub, &proc)
      end
    end

    def subscribe_to_all_events(subscriber, &proc)
      @event_broker.add_global_subscriber(subscriber).tap do |unsub|
        handle_subscribe(unsub, &proc)
      end
    end

    private

    def handle_subscribe(unsub, &proc)
      begin
        proc.call
      ensure
        unsub.()
      end if proc
    end
end
```

or make PubSub::Broker respond to (publish_event, subscribe, subscribe_to_all_events).

Either solution would make it possible to substitute it instead of RES::Client and use with the same API.

What do you think @andrzejsliwa @mlomnicki @mpraglowski @pawelpacana @fidel @sevos @siemakuba @vbyno ?

**Comment by siemakuba** [comment]
Sorry for being quiet that long ;)
Using the Broker directly was my first thought, but as you mentioned - the subscription handling logic would still enforce having the Client as well. The wrapper idea is interesting but in my opinion, it introduces quite a lot of overhead as compared to having a client configured with a simple NullRepository object as a repository. 
We implemented it exactly this way, and it solved our issues perfectly. 

I've seen that the issue we reported (#188) has been resolved and is ready to release - we will look into that, but probably would stick with the null object repo for a while.

Saying all that, there's nothing but closing the PR that is left ;) 
Thank you for your comments and support!

**Pull request closed without merging**